### ISSUE TITLE
Question about computing losses

### ISSUE THREAD

**Issue by buchholzmd** [daa233/generative-inpainting-pytorch#29]
Hi,

Thank you for your work, it is very helpful. 

I was wondering if you could explain why you use `x2_inpaint.detach()` when running a forward pass through the discriminator for calculating the discriminator's loss, but do not call `detach()` when calling the discriminator for calculating the generator's loss.

In this case, gradient will not be computed for the D loss for the fake images, but will for the G loss.

I understand why you use `detach()` when calculating the gradient penalty (as you only want to use the gradient for the GP term and not recalculate gradient for the discriminator again, but I cannot seem to understand why it is used as described above.

Thank you

**Comment by daa233** [comment]
Hi,

For the `detach()` here, I followed the [official PyTorch DCGAN implementation example](https://github.com/pytorch/examples/blob/master/dcgan/main.py). 

Here is the discussion that may help you: [why is detach necessary](https://github.com/pytorch/examples/issues/116)

In my opinion, it is mainly for efficiency and not retaining the graph.

**Comment by buchholzmd** [comment]
@daa233 Thank you very much for the quick response! I had a feeling it was purely for efficiency, it seems it is just to avoid recalculating the gradient for the inpainted  images. Both links you provided are very helpful.

Thank you again for your work!
